core/src/main/java/ai/timefold/solver/core/api/move/Move.java

```diff
@@ -0,0 +1,138 @@
+     *        Remembers those mutative operations and can replay them in reverse order
+     *        when the solver needs to undo the move.
+     *        Do not store this parameter in a field.
+     */
+    void execute(MutableSolutionState<Solution_> mutableSolutionState);
+
+    /**
+     * Rebases a move from an origin {@link ScoreDirector} to another destination {@link ScoreDirector}
+     * which is usually on another {@link Thread}.
+     * It is necessary for multithreaded solving to function.
+     * <p>
+     * The new move returned by this method translates the entities and problem facts
+     * to the destination {@link PlanningSolution} of the destination {@link ScoreDirector},
+     * That destination {@link PlanningSolution} is a deep planning clone (or an even deeper clone)
+     * of the origin {@link PlanningSolution} that this move has been generated from.
+     * <p>
+     * That new move does the exact same change as this move,
+     * resulting in the same {@link PlanningSolution} state,
+     * presuming that destination {@link PlanningSolution} was in the same state
+     * as the original {@link PlanningSolution} to begin with.
+     * <p>
+     * An implementation of this method typically iterates through every entity and fact instance in this move,
+     * translates each one to the destination {@link ScoreDirector} with {@link Rebaser#rebase(Object)}
+     * and creates a new move instance of the same move type, using those translated instances.
+     * <p>
+     * The destination {@link PlanningSolution} can be in a different state than the original {@link PlanningSolution}.
+     * So, rebasing can only depend on the identity of {@link PlanningEntity planning entities}
+     * and {@link ProblemFactProperty problem facts},
+     * which are usually declared by a {@link PlanningId} on those classes.
+     * It must not depend on the state of the {@link PlanningVariable planning variables}.
+     * One thread might rebase a move before, amid or after another thread does that same move instance.
+     * <p>
+     * This method is thread-safe.
+     *
+     * @param rebaser never null; do not store this parameter in a field
+     * @return never null, a new move that does the same change as this move on another solution instance
+     */
+    Move<Solution_> rebase(Rebaser rebaser);
+
+    /**
+     * Returns all planning entities that this move is changing.
+     * Required for entity tabu.
+     * <p>
+     * This method is only called after {@link #execute(MutableSolutionState)}, which might affect the return values.
+     * <p>
+     * Duplicate entries in the returned {@link Collection} are best avoided.
+     * The returned {@link Collection} is recommended to be in a stable order.
+     * For example, use {@link List} or {@link LinkedHashSet}, but not {@link HashSet}.
+     *
+     * @return never null
+     */
+    default Collection<?> extractPlanningEntities() {
```

Comment:
Are there any move implementations that do not provide these methods? Does it make sense to have them for any move? Depending on the answers, why not transfer these tabu-related operations to a specific tabu-move interface, such as `TabuMove extends Move`?

Next to that, the TB logic should use `TabuMove` instead of
`Move`. Does it make sense?

Reply:
Interesting idea. Either is good, IMO. In order to not complicate the type hierarchy, I don't see a problem with keeping it like this.